### PR TITLE
melange-decoders: melange backend

### LINKED ISSUE #59: Install bs-decoders from opam
Hi,

Please can you make bs-decoders installable from opam. It would be easier for beginners like me to use them with [Melange](https://github.com/melange-re/melange).

I did eventually figure out how to use bs-decoders (installed from npm) using stanzas, but I also had to copy `shims_let_ops_.ml` in `node_modules/bs-decoders/src-bs` into the `node_modules/bs-decoders/src` folder.

I did see:
```
(rule
 (targets shims_let_ops_.ml)
 (action
  (with-stdout-to
   %{targets}
   (run ./gen/mkshims.exe))))
```
but this is beyond a beginner like me.

I can try making a pull request if you can point me in the right direction as to how to make bs-decoders installable from opam.

Thanks.

### PR DESCRIPTION
**Pull request by actionshrimp**
- Combine bucklescript and melange implementations under `js`.
- This is breaking for bucklescript users: They now have to e.g. `open Decoders.Js_json` instead of `Decoders.Bs_json`

When finished this should fix #59.

TODO:
- [x] update build to run the tests for the melange version
- [ ] try consuming the lib from another project via `opam pin` to check the `install` hack works properly

### CODE REVIEW

**Review by mattjbray — APPROVED** [review]

File: dune-project [hunk 1]
```diff
@@ -127,6 +130,18 @@
...
   (ounit2 :with-test)))
 
+(package
+ (name decoders-mel)
+ (synopsis "Melange backend for decoders")
+ (description
+   "A combinator library for \"decoding\" JSON-like values into your own Ocaml types, inspired by Elm's `Json.Decode` and `Json.Encode`.")
+ (depends
+  (ocaml (>= 4.03.0))
+  (decoders (= :version))
```
Comment (mattjbray) [inline dune-project:140]: Is this needed, as `src-js/dune` copies in the source files anyway?

**Comment by mattjbray** [comment]
Is it possible to put all the js-related stuff under a `melange/` subfolder? Something like:

* Move `src-js` -> `melange/src`
* Create `melange/dune-project` with `(lang dune 3.8)` and `(using melange)` (and revert `/dune-project`)
* Dune will create `melange/decoders-mel.opam`
* `melange/src/dune` does the `(copy_files# ../../src/*.ml{,i})`

**Comment by mattjbray** [comment]
> Is it possible to put all the js-related stuff under a `melange/` subfolder? Something like:

(And we'd need to figure out how to keep the bucklescript build working under this new layout too)

File: js/src-js/shims_let_ops_.ml
Comment (mattjbray) [inline js/src-js/shims_let_ops_.ml:1]: Maybe we can copy this rule to generate this file https://github.com/mattjbray/ocaml-decoders/blob/62dc7f8c73919f505adc2f5b7374e8e5ea21f98e/src/dune#L7-L12

**Review by mattjbray — COMMENTED** [review]
Nice.

File: js/src-js/shims_let_ops_.ml
Comment (actionshrimp) [reply js/src-js/shims_let_ops_.ml:1]: I did try that - it didn't know about the `.exe` as it's not part of the same `dune-project` unfortunately.

**Comment by actionshrimp** [comment]
Naming the melange package `melange-decoders` (as opposed to `decoders-mel`) to match the melange convention described here: https://github.com/melange-re/melange/discussions/688

**Comment by actionshrimp** [comment]
Turns out we actually need `(modes melange ..)` for the `decoders` package itself in order to be able to use the Decoders sig in shared code - the shared code dune files will also need `(modes melange ..)` and `(libraries decoders ..)` for that to work.

So it looks like the `melange` dep  and dune version bump is unavoidable. One temporary solution could be to leave this melange compatible version of decoders on a branch and rebase it off the main version. That way people who need `melange` support can pin the branch, and other people who don't use melange can use the regular version?

**Comment by actionshrimp** [comment]
Superceded by #65

**Pull request closed without merging**